rivertype/river_type.go

```diff
@@ -218,6 +219,128 @@ type AttemptError struct {
+	MaxAttempts int
+	Metadata    []byte
+	Priority    int
+	Queue       string
+	ScheduledAt *time.Time
+	State       JobState
+	Tags        []string
+}
+
+// JobMiddleware provides an interface for middleware that integrations can use
+// to encapsulate common logic around various phases of a job's lifecycle.
+//
+// Implementations should embed river.JobMiddlewareDefaults to inherit default
+// implementations for phases where no custom code is needed, and for forward
+// compatibility in case new functions are added to this interface.
+type JobMiddleware interface {
+	// Insert is invoked around an insert operation. Implementations must always
+	// include a call to doInner to call down the middleware stack and perfom
+	// the insertion, and may run custom code before and after.
+	//
+	// Returning an error from this function will fail the overarching insert
+	// operation, even if the inner insertion originally succeeded.
+	//
+	// Insert is *not* invoked on batch insertions using Client.InsertMany or
+	// Client.InsertManyTx. InsertMany should be implemented separately.
+	Insert(ctx context.Context, params *JobInsertParams, doInner func(ctx context.Context) (*JobInsertResult, error)) (*JobInsertResult, error)
+
+	// InsertMany is invoked around a batch insert operation. Implementations
+	// must always include a call to doInner to call down the middleware stack
+	// and perfom the batch insertion, and may run custom code before and after.
+	//
+	// Returning an error from this function will fail the overarching insert
+	// operation, even if the inner insertion originally succeeded.
+	InsertMany(ctx context.Context, manyParams []*JobInsertParams, doInner func(ctx context.Context) (int, error)) (int, error)
+
+	// Work is invoked around a job's JSON args being unmarshaled and the job
+	// worked. Implementations must always include a call to doInner to call
+	// down the middleware stack and perfom the batch insertion, and may run
+	// custom code before and after.
+	//
+	// Returning an error from this function will fail the overarching work
+	// operation, even if the inner work originally succeeded.
+	Work(ctx context.Context, job *JobRow, doInner func(ctx context.Context) error) error
+
+	//
+	// below this line was an experiment and commented out now
+	//
+
+	// InsertBegin is invoked immediately before doing a job uniqueness check
+	// and (provided there was no duplicate), inserting it to the database. A
+	// lifecycle hook may make modifications to insert params, but should be
+	// careful doing so as it'll fundamentally modify the job.
+	//
+	// Returns a context that'll be used during the insert operation. Hooks that
+	// don't need to modify context should pass through the input ctx.
+	//
+	// InsertBegin is *not* invoked on a batch insertion with InsertMany or
+	// InsertManyTx. Integrations should implement InsertManyBegin separately.
+	// InsertBegin(ctx context.Context, params *JobLifecycleInsertParams) (context.Context, error)
```

Comment:
So I originally tried to write this system as "hooks" instead of middleware, and you can see what the old entrypoints looked like here (left them in as gravestones for the time being). They almost worked, but I found that I was running into two major problems with them compared to middleware:

* It'd be a common thing to want to be put something in context in the "begin" part, then extract that thing in the "end" part (e.g. for metrics or whatever). I accomplished that by returning a context here as you can see, but it was fairly awkward, and I don't think users would've like it.
* The bigger issue is that it was hard to know how the "end" functions should be called (or not called) under various error conditions like a panic or error returned. We would've either need to send back parameters for all of (1) successful return, (2) possible error, (3) panic val, OR just have totally separate functions for when errors occurred, but both those options were extremely awkward. With middleware, the return values are right there, and it's up to the caller to just do whatever they want with them.

Reply:
Makes sense, I think this is probably the cleaner & more flexible option.